core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -264,29 +297,12 @@ public void handleException(TransportException exp) {
                                 super.handleException(exp);
                             }
                         });
-                } else {
-                    setPhase(replicationTask, "primary");
-                    final IndexMetaData indexMetaData = clusterService.state().getMetaData().index(request.shardId().getIndex());
-                    final boolean executeOnReplicas = (indexMetaData == null) || shouldExecuteReplication(indexMetaData.getSettings());
-                    final ActionListener<Response> listener = createResponseListener(channel, replicationTask, primaryShardReference);
-                    createReplicatedOperation(request, new ActionListener<PrimaryResult>() {
-                        @Override
-                        public void onResponse(PrimaryResult result) {
-                            result.respond(listener);
-                        }
-
-                        @Override
-                        public void onFailure(Throwable e) {
-                            listener.onFailure(e);
-                        }
-                    }, primaryShardReference, executeOnReplicas).execute();
-                    success = true;
-                }
-            } finally {
-                if (success == false) {
-                    primaryShardReference.close();
+                } catch (Throwable t) {
+                    onFailure.handle(t);
                 }
-            }
+            };
+
+            acquirePrimaryShardReference(request.shardId(), onReferenceAcquired, onRelocated, onFailure);
```

Comment:
why can't we use an ActionListener<PrimaryShardReference> call back here? I think we can deal with the relocation check in the same way did before, within the onResponse method?